src/dynamics/mpas/dp_coupling.F90

```diff
@@ -585,30 +776,132 @@ subroutine hydrostatic_pressure(nCells, nVertLevels, zz, zgrid, rho_zz, theta_m,
         
         thetavk = theta_m(k,iCell)/ (1.0_r8 + q(k,iCell)) !convert modified theta to virtual theta
-        pmid(k,iCell) = (rhok*rgas*thetavk*(kap1))**kap2  !mid-level pressure
+        pmid(k,iCell) = (rhok*rgas*thetavk*kap1)**kap2    !mid-level pressure
       end do
     end do
 end subroutine hydrostatic_pressure
 
+
+subroutine tot_energy(nCells, nVertLevels, qsize, index_qv, zz, zgrid, rho_zz, theta_m, q, ux,uy,outfld_name_suffix)
+  use physconst,      only: rair, cpair, gravit,cappa!=R/cp (dry air)  
+  use physconst,      only: thermodynamic_active_species_liq_num
+  use mpas_constants, only: p0,cv,rv,rgas,cp
+  use cam_history,    only: outfld, hist_fld_active
+  use mpas_constants, only: Rv_over_Rd => rvord
+  use physconst,      only: thermodynamic_active_species_ice_idx_dycore,thermodynamic_active_species_liq_idx_dycore
+  use physconst,      only: thermodynamic_active_species_ice_num,thermodynamic_active_species_liq_num
+  ! Arguments
+  integer, intent(in) :: nCells
+  integer, intent(in) :: nVertLevels
+  integer, intent(in) :: qsize
+  integer, intent(in) :: index_qv
+  real(r8), dimension(nVertLevels, nCells),       intent(in) :: zz      ! d(zeta)/dz [-]
+  real(r8), dimension(nVertLevels+1, nCells),     intent(in) :: zgrid   ! geometric heights of layer interfaces [m]
+  real(r8), dimension(nVertLevels, nCells),       intent(in) :: rho_zz  ! dry density / zz [kg m^-3]
+  real(r8), dimension(nVertLevels, nCells),       intent(in) :: theta_m ! modified potential temperature
+  real(r8), dimension(qsize,nVertLevels, nCells), intent(in) :: q       ! tracer array
+  real(r8), dimension(nVertLevels, nCells),       intent(in) :: ux      ! A-grid zonal velocity component
+  real(r8), dimension(nVertLevels, nCells),       intent(in) :: uy      ! A-grid meridional velocity component
+  character*(*),                                  intent(in) :: outfld_name_suffix ! suffix for "outfld" names
+
+  ! Local variables
+  integer :: iCell, k, idx
+  real(r8) :: rho_dz,dz,zcell,temperature,rhod,theta,pk,ptop,exner
+  real(r8), dimension(nCells):: kinetic_energy,potential_energy,internal_energy,water_vapor,water_liq,water_ice
+  
+  real(r8), dimension(nCells) :: liq !total column integrated liquid
+  real(r8), dimension(nCells) :: ice !total column integrated ice
+
+  character(len=16) :: name_out1,name_out2,name_out3,name_out4,name_out5,name_out6
+  
+  name_out1 = 'SE_'   //trim(outfld_name_suffix)
+  name_out2 = 'KE_'   //trim(outfld_name_suffix)
+  name_out3 = 'WV_'   //trim(outfld_name_suffix)
+  name_out4 = 'WL_'   //trim(outfld_name_suffix)
+  name_out5 = 'WI_'   //trim(outfld_name_suffix)
+  name_out6 = 'TT_'   //trim(outfld_name_suffix)   
+  
+  if ( hist_fld_active(name_out1).or.hist_fld_active(name_out2).or.hist_fld_active(name_out3).or.&
+       hist_fld_active(name_out4).or.hist_fld_active(name_out5).or.hist_fld_active(name_out6)) then
```

Comment:
The if-statements are in there so that we do not compute the vertical integrals if we are not going to output them. The way the subroutine is used in dp_coupling makes the logicals not necessary (as you are pointing out).

There is a similar subroutine in physics (calc_te_and_aam_budgets) and the calls to that subroutine do not have logicals around them in which case we would be computing several vertical integrals even if the energy diagnostics are not active. That is where I got the code from.

I am fine with either approach